source/fab/tasks/c.py

```diff
@@ -0,0 +1,535 @@
+        previous_id = None
+        info: Optional[CInfo] = None
+        for row in rows:
+            symbol_id = CSymbolID(row['symbol'], Path(row['found_in']))
+            if previous_id is not None and symbol_id == previous_id:
+                if info is not None:
+                    info.add_prerequisite(row['prerequisite'])
+            else:  # symbol_id != previous_id
+                if info is not None:
+                    info_list.append(info)
+                info = CInfo(symbol_id)
+                if row['prerequisite'] is not None:
+                    info.add_prerequisite((row['prerequisite']))
+                previous_id = symbol_id
+        if info is not None:  # We have left overs
+            info_list.append(info)
+        if len(info_list) == 0:
+            message = 'symbol "{symbol}" not found in database.'
+            raise WorkingStateException(message.format(symbol=name))
+        return info_list
+
+    def depends_on(self, symbol: CSymbolID)\
+            -> Generator[CSymbolID, None, None]:
+        """
+        Gets the prerequisite symbols of a symbol.
+        :param symbol: symbol identifier.
+        :return: Prerequisite symbol names. May be an empty list.
+        """
+        query = '''select p.prerequisite, f.found_in
+                   from c_prerequisite as p
+                   left join c_symbol as f on f.symbol = p.prerequisite
+                   where p.symbol=:symbol and p.found_in=:filename
+                   order by p.symbol, f.found_in'''
+        rows = self.execute(query, {'symbol': symbol.name,
+                                    'filename': str(symbol.found_in)})
+        for row in rows:
+            if row['found_in'] is None:
+                yield CSymbolUnresolvedID(row['prerequisite'])
+            else:  # row['found_in'] is not None
+                yield CSymbolID(row['prerequisite'], Path(row['found_in']))
+
+
+class CAnalyser(Task):
+    def __init__(self, workspace: Path):
+        self.database = SqliteStateDatabase(workspace)
+
+    def _locate_include_regions(self, trans_unit) -> None:
+        # Aim is to identify where included (top level) regions
+        # start and end in the file
+        self._include_region = []
+
+        # Use a deque to implement a rolling window of 4 identifiers
+        # (enough to be sure we can spot an entire pragma)
+        identifiers: deque = deque([])
+        for token in trans_unit.cursor.get_tokens():
+            identifiers.append(token)
+            if len(identifiers) < 4:
+                continue
+            if len(identifiers) > 4:
+                identifiers.popleft()
+
+            # Trigger off of the FAB identifier only to save
+            # on joining the group too frequently
+            if identifiers[2].spelling == "FAB":
+                lineno = identifiers[2].location.line
+                full = " ".join(id.spelling for id in identifiers)
+                if full == "# pragma FAB SysIncludeStart":
```

Comment:
Whenever I've seen pragmas in use they have presented as `#pragma ...`, without the space. Please confirm that the space is optionally allowed.

Reply:
Yes you're right, it got lost in translation somewhere between my testing and implementing it properly

Reply:
Ah, no it didn't - it is a bit confusing but because that step on L292 above is doing `" ".join()` it is adding a space inbetween, but it's only there for the purposes of doing the check (not in the source)